docs/src/adr/018adr-inlining.md

```diff
@@ -0,0 +1,73 @@
+# ADR-018: Inlining
+
+| author     | revision | last revised |
+| ------------ | --------:| -------- |
+| Jure Kukovec |    1 | Apr 13, 2022 |
+
+**Table of Contents**
+
+- [Summary (Overview)](#summary)
+- [Context (Problem)](#context)
+- [Options (Alternatives)](#options)
+- [Solution (Decision)](#solution)
+- [Consequences (Retrospection)](#consequences)
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+This ADR outlines the transformations that happen in the Apalache inliner pass.
+Since we have recently reworked the inliner in #1569, we saw it fit to document exactly how inlining is supposed to work and why the inlining pass is the way it is. 
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+The term "inlining" typically refers to the process of replacing instances of operator application `A(x1,...,xn)`, with `e[x1/p1,...,xn/pn]` - the expression obtained by replacing each instance of `pi` with `xi` -  where the operator `A` is defined as `A(p1,...,pn) == e` and is _non-recursive_.
+We elect to use the term in a broader sense of "replacing an operator with its definition", in the two ways listed below:
+  1. Standard inlining: the instantiation described above
+      1. Non-nullary inlining: the instantiation described above, except the inlining skips nullary LET-defined operators
+  2. Pass-by-name inlining: replacing an operator name `A` with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
+
+The reason for doing (1) is that, in order to encode `A(x1,...,xn)`, we need to know `e[x1/p1,...,xn/pn]`, not just `e`, since the latter may contain free variables (e.g. `p1` is free in `e`).
+
+The reason for doing (2) is more pragmatic; In order to rewrite expressions which feature any of the higher-order (HO) built-in operators, e.g. `ApaFoldSet(A, v, S)`, we need to know, at the time of rewriting, how to evaluate an application of `A` (e.g. `A(partial, current)` for folding). 
+Performing (2) allows us to make the rewriting rule local, since the definition becomes available where the operator is used, and frees us from having to track scope in the rewriting rules.
+
+## Options
+
+<!-- Communicate the options considered.
+     This records evidence of our circumspection and documents the various alternatives
+     considered but not adopted.
+-->
+1. Perform no inlining in preprocessing and inline only as needed in the rewriting rules.
```

Comment:
Based on the document so far, I was unprepared that this will actually discuss different options.

Reply:
I agree. I think it's partially because the summary and context don't present the problem to be solved, which would invite thinking about possible solutions, but instead frame the document as an explanation of a technique we are already using.

Reply:
The thing is though, inlining isn't really a solution to some theoretical problem, it's a case of "this makes rewriting rules much more convenient".

Reply:
I suppose most real-world ADRs address practical problems, not theoretical ones.

> "this makes rewriting rules much more convenient"

Right. Any why this is a problem we're trying to solve, and what forces an effective solution must take into account, should be discussed as context for the reader.